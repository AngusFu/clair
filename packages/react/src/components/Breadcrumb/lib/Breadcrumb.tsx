import React from "react";
import PropTypes from "prop-types";
import { BreadcrumbContext } from "./Context";

type Text = React.ReactNode;
const PText = PropTypes.node;

export interface BreadcrumbProps {
  children: Text;
  separator?: React.ReactNode;
}

export const Breadcrumb: React.FC<BreadcrumbProps> = props => {
  const { children, separator, ...rest } = props;
  const contextValue = {
    separator
  };

  return (
    <nav className="c-breadcrumb" role="navigation" {...rest}>
      <BreadcrumbContext.Provider value={contextValue}>
        {children}
      </BreadcrumbContext.Provider>
    </nav>
  );
};

Breadcrumb.propTypes = {
  children: PText.isRequired,
  separator: PropTypes.node
};

Breadcrumb.displayName = "Breadcrumb";
