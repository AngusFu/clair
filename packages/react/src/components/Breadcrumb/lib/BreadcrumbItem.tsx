import React, { useContext } from "react";
import PropTypes from "prop-types";
import { BreadcrumbContext } from "./Context";

function hasTag(node: React.ReactNode, tag = "a") {
  const elem = node as React.ReactElement;
  if (!elem) return false;
  if (elem.type === tag) return true;
  if (!elem.props) return false;
  if (!Array.isArray(elem.props.children)) return false;
  return elem.props.children.some((child: React.ReactElement) =>
    hasTag(child, tag)
  );
}

export interface BreadcrumbItemProps
  extends React.HTMLAttributes<HTMLSpanElement> {
  href?: string;
  children?: React.ReactNode;
}

export const BreadcrumbItem: React.FC<BreadcrumbItemProps> = props => {
  const { children, href, ...rest } = props;
  const context = useContext(BreadcrumbContext);
  const separator = context.separator || "/";
  const hasATag = Array.isArray(children)
    ? children.some(c => hasTag(c))
    : hasTag(children);
  const tabIndex = href && !hasATag ? 0 : undefined;
  const role = href || hasATag ? "link" : undefined;
  const link = href ? <a href={href}>{children}</a> : children;

  return (
    <span className="c-breadcrumb__item">
      <span role={role} tabIndex={tabIndex} {...rest}>
        {link}
      </span>
      <span className="c-breadcrumb__separator">{separator}</span>
    </span>
  );
};

BreadcrumbItem.propTypes = {
  href: PropTypes.string,
  children: PropTypes.node
};

BreadcrumbItem.displayName = "BreadcrumbItem";
