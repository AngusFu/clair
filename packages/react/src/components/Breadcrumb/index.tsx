import { Breadcrumb } from "./lib/Breadcrumb";
import { BreadcrumbItem } from "./lib/BreadcrumbItem";

export { Breadcrumb, BreadcrumbItem };
