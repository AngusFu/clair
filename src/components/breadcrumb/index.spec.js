import { mount, shallowMount } from '@vue/test-utils'
import Breadcrumb from './index.vue'
import BreadcrumbItem from './breadcrumb-item.vue'

const components = {
  'c-breadcrumb': Breadcrumb,
  'c-breadcrumb-item': BreadcrumbItem
}

describe('breadcrumb-item', function () {
  it('should works', () => {
    const wrapper = shallowMount(BreadcrumbItem, {
      slots: { default: '<div>Content</div>' }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.text()).toContain('Content')
  })
})

describe('breadcrumb', function () {
  it('should works', () => {
    const template = `<c-breadcrumb>
  <c-breadcrumb-item><a href="#">首页</a></c-breadcrumb-item>
  <c-breadcrumb-item><a href="#">个人信息</a></c-breadcrumb-item>
  <c-breadcrumb-item>修改密码</c-breadcrumb-item>
</c-breadcrumb>`

    const wrapper = mount({
      template,
      components
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('works with custom divider', () => {
    const template = `<c-breadcrumb divider="▹">
  <c-breadcrumb-item><a href="#">首页</a></c-breadcrumb-item>
  <c-breadcrumb-item><a href="#">个人信息</a></c-breadcrumb-item>
  <c-breadcrumb-item>修改密码</c-breadcrumb-item>
</c-breadcrumb>`

    const wrapper = mount({ template, components })
    expect(wrapper.findAll('i').length).toBe(2)
    expect(wrapper).toMatchSnapshot()
  })

  it('works with scoped slot as custom divider', () => {
    const template = `<c-breadcrumb>
  <template slot="divider" slot-scope="props">
    <span class="divider">$</span>
  </template>
  <c-breadcrumb-item><a href="#">首页</a></c-breadcrumb-item>
  <c-breadcrumb-item><a href="#">个人信息</a></c-breadcrumb-item>
  <c-breadcrumb-item>修改密码</c-breadcrumb-item>
</c-breadcrumb>`

    const wrapper = mount({ template, components })
    expect(wrapper.text()).toContain('$')
    expect(wrapper).toMatchSnapshot()
  })
})
