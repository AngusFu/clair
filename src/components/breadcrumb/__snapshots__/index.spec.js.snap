// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breadcrumb should works 1`] = `
<nav class="c-breadcrumb">
  <div class="c-breadcrumb__item"><a href="#">首页</a></div><i class="c-breadcrumb__divider">/</i>
  <div class="c-breadcrumb__item"><a href="#">个人信息</a></div><i class="c-breadcrumb__divider">/</i>
  <div class="c-breadcrumb__item">修改密码</div>
</nav>
`;

exports[`breadcrumb works with custom divider 1`] = `
<nav class="c-breadcrumb">
  <div class="c-breadcrumb__item"><a href="#">首页</a></div><i class="c-breadcrumb__divider">▹</i>
  <div class="c-breadcrumb__item"><a href="#">个人信息</a></div><i class="c-breadcrumb__divider">▹</i>
  <div class="c-breadcrumb__item">修改密码</div>
</nav>
`;

exports[`breadcrumb works with scoped slot as custom divider 1`] = `
<nav class="c-breadcrumb">
  <div class="c-breadcrumb__item"><a href="#">首页</a></div><i class="c-breadcrumb__divider"><span class="divider">$</span></i>
  <div class="c-breadcrumb__item"><a href="#">个人信息</a></div><i class="c-breadcrumb__divider"><span class="divider">$</span></i>
  <div class="c-breadcrumb__item">修改密码</div>
</nav>
`;

exports[`breadcrumb-item should works 1`] = `
<div class="c-breadcrumb__item">
  <div>Content</div>
</div>
`;
